src/node/jsonrpc.rs

```diff
@@ -0,0 +1,149 @@
+use crate::node::config::{
+    IPCAgentJsonRPCNodeConfig, DEFAULT_JSON_RPC_ENDPOINT, DEFAULT_JSON_RPC_VERSION,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::convert::Infallible;
+use warp::Filter;
+
+/// The IPC JSON RPC node that contains all the methods and handlers. The underlying implementation
+/// is using `warp`.
+///
+/// Note that currently only http json rpc is supported.
+///
+/// # Examples
+/// ```no_run
+/// use agent::node::config::IPCAgentJsonRPCNodeConfig;
+/// use agent::node::jsonrpc::IPCAgentJsonRPCNode;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let n = IPCAgentJsonRPCNode::new(IPCAgentJsonRPCNodeConfig::default());
+///     n.run().await;
+/// }
+/// ```
+pub struct IPCAgentJsonRPCNode {
+    config: IPCAgentJsonRPCNodeConfig,
+}
+
+impl IPCAgentJsonRPCNode {
+    pub fn new(config: IPCAgentJsonRPCNodeConfig) -> Self {
+        Self { config }
+    }
+
+    /// Runs the node in the current thread
+    pub async fn run(&self) {
+        log::info!("rpc node started at {:?}", self.config.addr());
+        warp::serve(json_rpc_filter()).run(self.config.addr()).await;
+    }
+}
+
+// Internal implementations
+
+/// Create the json_rpc filter
+fn json_rpc_filter() -> impl Filter<Extract = (warp::reply::Json,), Error = warp::Rejection> + Copy
+{
+    warp::post()
+        .and(warp::path(DEFAULT_JSON_RPC_ENDPOINT))
+        .and(warp::body::bytes())
+        .and_then(process)
+}
+
+/// The process method for the JSON RPC node.
+async fn process(bytes: bytes::Bytes) -> Result<warp::reply::Json, Infallible> {
+    log::debug!("received bytes = {:?}", bytes);
+
+    match serde_json::from_slice::<JSONRPCParam>(bytes.as_ref()) {
+        Ok(p) => {
+            let JSONRPCParam {
+                id,
+                method,
+                params,
+                jsonrpc,
+            } = p;
```

Comment:
Why can't the deserialization into json be part of the filter (the function above), and then this function exclusively handle the routing/handling of functions? 

I would move this logic of deserializing the body into json up to the filter, and I would make this function be the one that picks up the request and receives the request and tailors the response.

You can add a placeholder here or a sample `echo` handler here to showcase how that would be structured.

Reply:
The main reason is because this way I dont really have to write another warp middleware, which should terminate the execution if deserialization fails. I was thinking of handling the routing by creating another method in this function body, will that work?

Reply:
I think is cleaner to have the deserialization on the top filter and the routing in this function. That way we have a clear separation of roles. Why can't you terminate the execution in the execution if the deserialization fails?

Reply:
Nah, just need to write more code which probably it's a style preference and might not have liked. I have updated the code accordingly, check it out:)